src/github.com/couchbase/sync_gateway/base/bucket_gocb.go

```diff
@@ -93,91 +97,416 @@ func (bucket CouchbaseBucketGoCB) GetName() string {
+		return shouldRetry, err, uint64(casGoCB)
+
+	}
+
+	sleeper := CreateDoublingSleeperFunc(
+		bucket.spec.MaxNumRetries,
+		bucket.spec.InitialRetrySleepTimeMS,
+	)
+
+	// Kick off retry loop
+	description := fmt.Sprintf("Get %v", k)
+	err, result := RetryLoop(description, worker, sleeper)
+
+	// If the retry loop returned a nil result, set to 0 to prevent type assertion on nil error
+	if result == nil {
+		result = uint64(0)
+	}
+
+	// Type assertion of result
+	cas, ok := result.(uint64)
+	if !ok {
+		LogPanic("Error doing type assertion of %v into a uint64", result)
+	}
+
+	return cas, err
+
 }
 
-func (bucket CouchbaseBucketGoCB) GetRaw(k string) (rv []byte, cas uint64, err error) {
+// Retry up to the retry limit, then return.  Does not retry items if they had CAS failures,
+// and it's up to the caller to handle those.
+func (bucket CouchbaseBucketGoCB) SetBulk(entries []*sgbucket.BulkSetEntry) (err error) {
 
-	bucket.singleOps <- struct{}{}
-	gocbExpvars.Add("SingleOps", 1)
-	defer func() {
-		<-bucket.singleOps
-		gocbExpvars.Add("SingleOps", -1)
-	}()
+	// Create the RetryWorker for BulkSet op
+	worker := bucket.newSetBulkRetryWorker(entries)
 
-	var returnVal interface{}
+	// this is the function that will be called back by the RetryLoop to determine
+	// how long to sleep before retrying (uses backoff)
+	sleeper := CreateDoublingSleeperFunc(
+		bucket.spec.MaxNumRetries,
+		bucket.spec.InitialRetrySleepTimeMS,
+	)
+
+	// Kick off retry loop
+	description := fmt.Sprintf("SetBulk with %v entries", len(entries))
+	err, _ = RetryLoop(description, worker, sleeper)
+
+	return err
+
+}
+
+func (bucket CouchbaseBucketGoCB) newSetBulkRetryWorker(entries []*sgbucket.BulkSetEntry) RetryWorker {
+
+	pendingKeys := createPendingKeysFromEntries(entries)
+	entriesKeysToIndex := createEntriesKeyToIndex(entries)
+
+	worker := func() (shouldRetry bool, err error, value interface{}) {
+
+		retryKeys := []string{}
+
+		// break up into batches
+		pendingKeysBatches := createBatchesKeys(MaxBulkBatchSize, pendingKeys)
+
+		for _, pendingKeysBatch := range pendingKeysBatches {
+			err, retryKeysForBatch := bucket.processEntriesBatch(
+				pendingKeysBatch,
+				entries,
+				entriesKeysToIndex,
+			)
+			retryKeys = append(retryKeys, retryKeysForBatch...)
+			if err != nil {
+				return false, err, nil
+			}
+		}
+
+		// if there are no keys to retry, then we're done.
+		if len(retryKeys) == 0 {
+			return false, nil, nil
+		}
+
+		// otherwise, retry the keys that need to be retried
+		pendingKeys = retryKeys
+
+		// return true to signal that this function needs to be retried
+		return true, nil, nil
 
-	gocbExpvars.Add("GetRaw", 1)
-	casGoCB, err := bucket.Bucket.Get(k, &returnVal)
-	if err != nil {
-		return nil, 0, err
 	}
-	return returnVal.([]byte), uint64(casGoCB), err
+
+	return worker
+
+}
+
+func (bucket CouchbaseBucketGoCB) processEntriesBatch(pendingKeys []string, entries []*sgbucket.BulkSetEntry, keyToIndex map[string]int) (error, []string) {
```

Comment:
Maybe rename to make it clear that this is processing a batch of set operations, compared to processBatch below?